internal/controllers/nmc_reconciler.go

```diff
@@ -520,56 +524,101 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
-	nodeModuleReadyLabels := sets.New[string]()
+	nodeModuleReadyLabels := sets.New[types.NamespacedName]()
+
 	for label := range node.GetLabels() {
-		if utils.IsKernelModuleReadyNodeLabel(label) {
-			nodeModuleReadyLabels.Insert(label)
+		if ok, namespace, name := utils.IsKernelModuleReadyNodeLabel(label); ok {
+			nodeModuleReadyLabels.Insert(types.NamespacedName{Namespace: namespace, Name: name})
 		}
 	}
 
 	// get spec labels and their config
-	specLabels := make(map[string]kmmv1beta1.ModuleConfig)
+	specLabels := make(map[types.NamespacedName]kmmv1beta1.ModuleConfig)
 	for _, module := range nmc.Spec.Modules {
-		label := utils.GetKernelModuleReadyNodeLabel(module.Namespace, module.Name)
-		specLabels[label] = module.Config
+		specLabels[types.NamespacedName{Namespace: module.Namespace, Name: module.Name}] = module.Config
 	}
 
 	// get status labels and their config
-	statusLabels := make(map[string]kmmv1beta1.ModuleConfig)
+	statusLabels := make(map[types.NamespacedName]kmmv1beta1.ModuleConfig)
 	for _, module := range nmc.Status.Modules {
-		label := utils.GetKernelModuleReadyNodeLabel(module.Namespace, module.Name)
+		label := types.NamespacedName{Namespace: module.Namespace, Name: module.Name}
 		statusLabels[label] = kmmv1beta1.ModuleConfig{}
 		if module.Config != nil {
 			statusLabels[label] = *module.Config
 		}
 	}
 
-	patchFrom := client.MergeFrom(node.DeepCopy())
+	unloaded := make([]types.NamespacedName, 0, len(nodeModuleReadyLabels))
+	loaded := make([]types.NamespacedName, 0, len(specLabels))
+
 	// label in node but not in spec or status - should be removed
-	for label := range nodeModuleReadyLabels {
-		_, inSpec := specLabels[label]
-		_, inStatus := statusLabels[label]
+	for nsn := range nodeModuleReadyLabels {
+		_, inSpec := specLabels[nsn]
+		_, inStatus := statusLabels[nsn]
 		if !inSpec && !inStatus {
-			meta.RemoveLabel(&node, label)
+			unloaded = append(unloaded, nsn)
 		}
 	}
 
 	// label in spec and status and config equal - should be added
-	for label, specConfig := range specLabels {
-		statusConfig, ok := statusLabels[label]
-		if ok && reflect.DeepEqual(specConfig, statusConfig) {
-			meta.SetLabel(&node, label, "")
+	for nsn, specConfig := range specLabels {
+		statusConfig, ok := statusLabels[nsn]
+		if ok && reflect.DeepEqual(specConfig, statusConfig) && !nodeModuleReadyLabels.Has(nsn) {
+			loaded = append(loaded, nsn)
 		}
 	}
 
-	return h.client.Patch(ctx, &node, patchFrom)
+	patchFrom := client.MergeFrom(node.DeepCopy())
+
+	for _, nsn := range unloaded {
+		meta.RemoveLabel(
+			&node,
+			utils.GetKernelModuleReadyNodeLabel(nsn.Namespace, nsn.Name),
+		)
+	}
+
+	for _, nsn := range loaded {
+		meta.SetLabel(
+			&node,
+			utils.GetKernelModuleReadyNodeLabel(nsn.Namespace, nsn.Name),
+			"",
+		)
+	}
+
+	if err := h.client.Patch(ctx, &node, patchFrom); err != nil {
+		return fmt.Errorf("could not patch node: %v", err)
+	}
+
+	for _, nsn := range unloaded {
```

Comment:
a couple of  general questions: 
1) AnnotatedEventF is a void function ( no error is returned). So why not just call the recorder in the SyncStatus, where we know exactly that Load/Unload worker pods have finished successfully?
2)  For build event we create events in a separate controller asynchronously, while for load/unload events we do it synchronously  in the NMC controller.  I think we should decide on a unified way to do it

Reply:
> AnnotatedEventF is a void function ( no error is returned). So why not just call the recorder in the SyncStatus, where we know exactly that Load/Unload worker pods have finished successfully?

That is another option. If we moved event recording to `SyncStatus`, we would still have to create two slices for loaded / unloaded, since we do not want to send the events before the status is updated (to avoid sending duplicate events when the status update fails). Please let me know what you prefer.

> For build event we create events in a separate controller asynchronously, while for load/unload events we do it synchronously in the NMC controller. I think we should decide on a unified way to do it

We could also send these in a separate controller. I thought it would be simpler to do it here in that function that updates the labels, because it is kind of the same thing. Please let me know what you prefer.